File: leetcode/912.java
import static java.lang.Math.min;

final class Solution {
    private static int set(
        final int[] array,
        final int idx,
        final int value
    ) {
        final var previous = array[idx];
        array[idx] = value;
        return previous;
    }

    private static void rotate(
        final int[] array,
        final int from,
        final int to,
        int distance
    ) {
        final var size = to - from;
        if (size == 0)
            return;
        distance = distance % size;
        if (distance == 0)
            return;
        if (distance < 0)
            distance += size;

        final var bound = from + size - distance;
        for (int cycleStart = from, nMoved = 0; nMoved < size; ++cycleStart) {
            var displaced = array[cycleStart];
            var i = cycleStart;
            do {
                if (i >= bound)
                    i -= size;
                i += distance;
                displaced = set(array, i, displaced);
                ++nMoved;
            } while (i != cycleStart);
        }
    }

    private static int lowerBound(final int[] array, final int key) {
        return lowerBound(array, 0, array.length, key);
    }

    private static int lowerBound(
        final int[] array,
        int fromIndex,
        int toIndex,
        final int key
    ) {
        --toIndex;
        while (fromIndex <= toIndex) {
            final var middle = fromIndex + toIndex >>> 1;
            if (array[middle] < key)
                fromIndex = middle + 1;
            else
                toIndex = middle - 1;
        }
        return fromIndex;
    }

    private static int upperBound(final int[] array, final int key) {
        return upperBound(array, 0, array.length, key);
    }

    public static <T> int upperBound(
        final int[] array,
        int fromIndex,
        int toIndex,
        final int key
    ) {
        --toIndex;
        while (fromIndex <= toIndex) {
            final var middle = fromIndex + toIndex >>> 1;
            if (array[middle] > key)
                toIndex = middle - 1;
            else
                fromIndex = middle + 1;
        }
        return fromIndex;
    }

    private static void merge(
        final int[] array,
        final int from,
        final int middle,
        final int to
    ) {
        final int leftLen = middle - from, rightLen = to - middle;
        if (leftLen == 0 || rightLen == 0)
            return;
        final int leftIdx, rightIdx, distance, newMiddle;
        if (leftLen < rightLen) {
            rightIdx = middle + (rightLen - 1 >> 1) + 1;
            leftIdx = upperBound(array, from, middle, array[rightIdx - 1]);
            distance = rightIdx - middle;
            newMiddle = leftIdx + distance - 1;
        } else {
            leftIdx = from + (leftLen - 1 >> 1);
            rightIdx = lowerBound(array, middle, to, array[leftIdx]);
            distance = rightIdx - middle;
            newMiddle = leftIdx + distance;
        }
        rotate(array, leftIdx, rightIdx, distance);
        merge(array, from, leftIdx, newMiddle);
        merge(array, newMiddle + 1, rightIdx, to);
    }

    public final int[] sortArray(final int[] nums) {
        final int length = nums.length, half = length >> 1;
        var width = 1;
        while (width < length) {
            final var newWidth = width <= half ? width << 1 : length;
            var i = 0;
            while (i < length) {
                final int bound = length - i, newI = i + min(newWidth, bound);
                merge(nums, i, i + min(width, bound), newI);
                i = newI;
            }
            width = newWidth;
        }
        return nums;
    }
}

File: leetcode/912.cpp
class Solution final {
private:
    template<class Iter>
    using Difference = iterator_traits<Iter>::difference_type;

    template<class Iter>
    using Value = iterator_traits<Iter>::value_type;

    template<class Iter>
    using Task = tuple<Iter, Iter, Iter>;

    template<class Iter>
    static constexpr auto
        category{typename iterator_traits<Iter>::iterator_category()};

    template<class T>
    static constexpr T midpointFloor(
        const T x,
        const T y
    ) noexcept {
        return (x & y) + ((x ^ y) >> 1);
    }

    template<class Iter>
    static constexpr pair<Iter, Iter> nthElement(
        Iter first1, Difference<Iter> length1,
        const Iter first2, const Difference<Iter> length2,
        Difference<Iter> n
    ) noexcept {
        if (n > length2) {
            first1 += n - length2;
            n = length2;
        }
        if (length1 > n)
            length1 = n;

        while (length1 > 0) {
            const auto index1{(length1 - 1) >> 1}, index2{n - index1 - 1};
            const auto iter1{next(first1, index1)}, iter2{next(first2, index2)};
            if (*iter2 < *iter1)
                length1 = index1;
            else {
                first1 = next(iter1);
                length1 -= index1 + 1;
                n -= index1 + 1;
            }
        }

        return {first1, next(first2, n)};
    }

    template<class Iter>
    static constexpr pair<Iter, Iter> nthElement(
        const Iter first1, const Iter last1,
        const Iter first2, const Iter last2,
        Difference<Iter> n
    ) noexcept {
        const auto
            length1{distance(first1, last1)}, length2{distance(first2, last2)};

        if (n <= 0)
            return {first1, first2};

        if (n - length1 >= length2)
            return {last1, last2};

        if (length1 > length2) {
            const auto [iter1, iter2]{
                nthElement(first2, length2, first1, length1, n)
            };
            return {iter2, iter1};
        }

        return nthElement(first1, length1, first2, length2, n);
    }

    template<class Iter, template<class> class Container = deque>
    class DefaultMerger final {
    private:
        queue<Value<Iter>, Container<Value<Iter>>> fifo{};

        constexpr void operator()(
            Iter first,
            const Iter middle,
            const Iter last,
            const forward_iterator_tag
        ) noexcept {
            if (first == middle || middle == last)
                return;

            auto second{middle};
            auto leftNonEmpty{true};
            do {
                if (leftNonEmpty)
                    fifo.push(move(*first));
                if (empty(fifo) || (second != last && *second < fifo.front())) {
                    *first = move(*second);
                    ++second;
                } else {
                    *first = move(fifo.front());
                    fifo.pop();
                }
                ++first;
                leftNonEmpty = leftNonEmpty && first != middle;
            } while (leftNonEmpty || !empty(fifo));
        }

        constexpr void operator()(
            Iter first,
            const Iter middle,
            const Iter last,
            const random_access_iterator_tag
        ) noexcept {
            if (first == middle || middle == last)
                return;

            auto second{middle};
            do {
                if (first < middle)
                    fifo.push(move(*first));
                if (empty(fifo) || (second != last && *second < fifo.front())) {
                    *first = move(*second);
                    ++second;
                } else {
                    *first = move(fifo.front());
                    fifo.pop();
                }
                ++first;
            } while (first < middle || !empty(fifo));
        }

    public:
        constexpr void operator()(
            Iter first,
            const Iter middle,
            const Iter last
        ) noexcept {
            (*this)(first, middle, last, category<Iter>);
        }
    };

    template<class Iter, template<class> class Container = vector>
    class RotateMerger final {
    private:
        stack<Task<Iter>, Container<Task<Iter>>> lifo{};

    public:
        constexpr void operator()(
            Iter first,
            Iter middle,
            Iter last
        ) noexcept {
            lifo.emplace(first, middle, last);
            do {
                tie(first, middle, last) = lifo.top();
                lifo.pop();
                if (first == middle || middle == last)
                    continue;

                const auto leftSize{distance(first, middle)},
                    rightSize{distance(middle, last)};
                Iter leftIter, rightIter;
                if (leftSize < rightSize) {
                    rightIter = next(middle, (rightSize + 1) >> 1);
                    leftIter = upper_bound(first, middle, *prev(rightIter));
                    middle = prev(rotate(leftIter, middle, rightIter));
                } else {
                    leftIter = next(first, (leftSize - 1) >> 1);
                    rightIter = lower_bound(middle, last, *leftIter);
                    middle = rotate(leftIter, middle, rightIter);
                }

                lifo.emplace(first, leftIter, middle);
                lifo.emplace(next(middle), rightIter, last);
            } while (!empty(lifo));
        }
    };

    template<class Iter, template<class> class Container = vector>
    class RotatePartitionMerger final {
    private:
        stack<Task<Iter>, Container<Task<Iter>>> lifo{};

    public:
        constexpr void operator()(
            Iter first,
            Iter middle,
            Iter last
        ) noexcept {
            lifo.emplace(first, middle, last);
            do {
                tie(first, middle, last) = lifo.top();
                lifo.pop();
                if (first == middle || middle == last)
                    continue;

                const auto n{
                    midpointFloor(distance(first, middle), distance(middle, last))
                };
                const auto [leftIter, rightIter]{
                    nthElement(first, middle, middle, last, n)
                };
                middle = rotate(leftIter, middle, rightIter);

                lifo.emplace(first, leftIter, middle);
                lifo.emplace(middle, rightIter, last);
            } while (!empty(lifo));
        }
    };

    template<class Iter, class Merger>
    static constexpr void mergeSort(
        const Iter first,
        const Iter last,
        Merger &&merger
    ) noexcept {
        const auto length{distance(first, last)}, half{length >> 1};
        Difference<Iter> step{1};
        while (step < length) {
            auto n{length};
            auto iter{first};
            while (n > 0) {
                const auto
                    left{min(step, n)},
                    right{min(step, n - left)};
                const auto middle{next(iter, left)},
                    nextIter{next(middle, right)};
                merger(iter, middle, nextIter);
                n -= left + right;
                iter = nextIter;
            }
            step = step <= half ? step << 1 : length;
        }
    }

    template<class Iter>
    static constexpr void defaultMergeSort(
        const Iter first,
        const Iter last
    ) noexcept {
        mergeSort(first, last, DefaultMerger<Iter>{});
    }

    template<class Iter>
    static constexpr void rotateMergeSort(
        const Iter first,
        const Iter last
    ) noexcept {
        mergeSort(first, last, RotateMerger<Iter>{});
    }

    template<class Iter>
    static constexpr void rotatePartitionMergeSort(
        const Iter first,
        const Iter last
    ) noexcept {
        mergeSort(first, last, RotatePartitionMerger<Iter>{});
    }

    template<class Iter>
    static constexpr void siftDown(
        const Iter first,
        Iter element,
        const Iter last
    ) noexcept {
        const auto half{next(first, distance(first, last) >> 1)};
        while (element < half) {
            auto child{next(element, distance(first, element) + 1)};
            if (const auto sibling{next(child)};
                sibling != last && *child < *sibling
            )
                child = sibling;
            if (*element >= *child)
                break;
            iter_swap(element, child);
            element = child;
        }
    }

    template<class Iter>
    static constexpr void siftUp(const Iter first, Iter element) noexcept {
        while (element != first) {
            const auto
                parent{prev(element, (distance(first, element) >> 1) + 1)};
            if (*parent >= *element)
                break;
            iter_swap(parent, element);
            element = parent;
        }
    }

    template<class Iter>
    static constexpr void popHeap(const Iter first, Iter last) noexcept {
        iter_swap(first, --last);
        siftDown(first, first, last);
    }

    template<class Iter>
    static constexpr void pushHeap(const Iter first, const Iter last) noexcept {
        siftUp(first, prev(last));
    }

    template<class Iter>
    static constexpr void makeHeap(const Iter first, const Iter last) noexcept {
        const auto length{distance(first, last)};
        if (length < 2)
            return;

        auto iter{next(first, length >> 1)};
        do
            siftDown(first, --iter, last);
        while (iter != first);
    }

    template<class Iter>
    static constexpr void sortHeap(const Iter first, Iter last) noexcept {
        if (distance(first, last) < 2)
            return;

        const auto second{next(first)};
        do
            popHeap(first, last);
        while (--last != second);
    }

    template<class Iter>
    static constexpr void heapSort(
        const Iter first,
        const Iter last
    ) noexcept {
        makeHeap(first, last);
        sortHeap(first, last);
    }

public:
    constexpr vector<int> sortArray(vector<int> &nums) const noexcept {
        heapSort(begin(nums), end(nums));
        return move(nums);
    }
};
